Create Withdrawals Reserves, central contract for money

gnidan opened the pull request:
Ref: #26 

This PR does a bunch of things, but mostly related to high level items:

- Fix the DollarToken tests!
- Define WithdrawalsReserves, to map participant balances and expose `reserve()` / `withdraw()` / `transfer()` functionality

gnidan committed:
Make deployService logs explicit

gnidan committed:
Remove unnecessary and duplicated Pending event

gnidan committed:
Add truffle-contract dependency

gnidan committed:
Add installed_contracts to gitignore

gnidan committed:
Add ExchangeRate price stubbing

gnidan committed:
Add internal function for runtime service lookup

gnidan committed:
Create initial WithdrawalsReserves

gnidan committed:
Add reservation interface to DollarToken

gnidan committed:
Hook up withdrawal reserves to factory

- Update emitContract()
- Update Factory tests

gnidan committed:
Hook up reservation on queue entry

gnidan committed:
Fix gas amounts

gnidan commented:
Recommend going commit-by-commit for this, might be easier to review.

gnidan commented:
Oh, question with this: what other tests might be worth adding?

gnidan commented:
Oh, also @chatch, I also removed the round-to-the-nearest dollar thing; was that there for a reason? It seemed like it made the math a lot trickier.

chatch commented:
On tests, what's there looks good, all new peices have tests. But it would be useful to update the "should handle full life cycle" in dollar_token.js with some checks on balances in the reserves and on the BVC state and things like that. I see this one as the one end to end test of the smart contracts side of things, it's an integration test really.  maybe it should go in it's own file even .. but yeah i reckon we keep that one updated and exercising one full cycle.

chatch commented:
A few points from review:

1. WithdrawalReserves.withdraw

Could just use Solidity send() instead of SafeSend. SafeSend is sending more gas then needed (10000 or 30000?) where send() will use just 2300. I think the extra checks it does aren't needed either as your already checking the balance and have previously checked (received from anyway) valid addresses as well, so there is a little more gas used to do these checks.

2. WithdrawalReserves.transfer

Does the BVC checkMargin need to get called on this initial deposit? Maybe deposit function could be split into 2 and have an initalDeposit that just accepts ether transfer. Then you could constrain the gas on the generic call to BVC deposit as well for a little more safety.  An initalDeposit function could further enforce msg.sender is WithdrawalReserves to block any other possible call to it. 

Maybe intentional because margins calculations aren't there yet but the first line in checkMargin is just "_;" so the rest of the logic gets ignored.

3. DollarToken reserveFor

could be marked private to block outside calls

4. Small gas save

Factory sha3("deposit()") change to constants at contract level.

chatch reviewed (changes requested):
See last comment with some potential changes.

gnidan commented:
Just one comment about this (will go through this for real later):

> But it would be useful to update the "should handle full life cycle" in dollar_token.js with some checks on balances in the reserves and on the BVC state and things like that.

There's [an outstanding bug](https://github.com/ethereumjs/testrpc/issues/122) with testrpc that might cause problems with testing for correct balance. I'll see if this will impact the integration tests, and if not, I'll make sure to include balance checking.

gnidan commented on contracts/WithdrawalsReserves.sol:
```diff
@@ -0,0 +1,79 @@
+        // re-entry protection
+        balances[participant] = 0;
+
+        uint sent = participant.safeSend(amount);
```
quoth @chatch:

> Could just use Solidity send() instead of SafeSend. SafeSend is sending more gas then needed (10000 or 30000?) where send() will use just 2300. I think the extra checks it does aren't needed either as your already checking the balance and have previously checked (received from anyway) valid addresses as well, so there is a little more gas used to do these checks.

gnidan replied:
(just so I can address these programmatically)

gnidan commented on contracts/WithdrawalsReserves.sol:
```diff
@@ -0,0 +1,79 @@
+        balances[_from] = newBalance;
+
+        Transfer(_from, _amount);
+    }
```
quoth @chatch:
> Does the BVC checkMargin need to get called on this initial deposit? Maybe deposit function could be split into 2 and have an initalDeposit that just accepts ether transfer. Then you could constrain the gas on the generic call to BVC deposit as well for a little more safety. An initalDeposit function could further enforce msg.sender is WithdrawalReserves to block any other possible call to it.
> 
> Maybe intentional because margins calculations aren't there yet but the first line in checkMargin is just "_;" so the rest of the logic gets ignored.

gnidan replied:
wait when you say
> the first line in checkMargin is just "_;" so the rest of the logic gets ignored.
do you mean to imply that logic doesn't run in Solidity after the underscore statement? that modifier _should_ be hooked up and performing margin calculation after deposit.

But I agree with your thought process and will split those up. The anti-pattern for this deposit() becomes apparent with this double transfer.

chatch replied:
Yes i was thinking the rest of the logic gets ignored. But that is wrong isn't it? It all gets executed after the function body ..

gnidan commented on contracts/DollarToken.sol:
```diff
@@ -86,10 +88,20 @@ contract DollarToken is Owned {
 
+    function reserveFor(address _participant)
+        payable
+    {
```
@chatch 
> could be marked private to block outside calls

Tried that - this doesn't work because of Solidity limitations with CALL vs. DELEGATECALL, I believe? I tried marking it `private` and `internal` and neither compile, I think because internal calls can't be payable.

But I am fuzzy on the details of this

chatch replied:
I think it should work if you drop payable from privates / internal. I was doing this with a previous version of the EntryChannel and I'm pretty sure it was working : **https://github.com/Bakt/matcher/blob/411966197cfbda38f90ba73e91eef5b8db77ce68/contracts/BidChannel.sol**
.

gnidan commented on contracts/Factory.sol:
```diff
@@ -16,15 +17,30 @@ contract Factory {
+        );
+
+        WithdrawalsReserves(reservesAddress).transfer(
+            participantWei, beneficiary, bvc, sha3("deposit()")
```
quoth @chatch:
> Factory sha3("deposit()") change to constants at contract level.

gnidan commented:
> On tests, what's there looks good, all new peices have tests. But it would be useful to update the "should handle full life cycle" in dollar_token.js with some checks on balances in the reserves and on the BVC state and things like that. I see this one as the one end to end test of the smart contracts side of things, it's an integration test really. maybe it should go in it's own file even .. but yeah i reckon we keep that one updated and exercising one full cycle.

Made https://github.com/Bakt/smart-contracts/issues/28 to reflect the work for splitting out the integration test

gnidan committed:
Standardize sha3(service) interface

- Move existing sha3 service name calculations into Services impl
  (for constant-time operation across-the-board)

- Expose abstract interface for each service in ServicesI

- Convert all external service invocations to this approach

gnidan committed:
Switch UPPER_CASE services to camelCase

gnidan committed:
Rename WithdrawalsReserves to WithdrawalReserves

gnidan committed:
Reduce gas cost of withdraw()

gnidan committed:
Make BVC store only reference to ServicesI

Instead of ExchangeRate (and others)

- Follow pattern for lookup-on-transaction (rather than consulting
  saved external address)

- Also, remove updateServices(), for reasons of needing requirements-spec

gnidan committed:
Refactor BVC deployment

- Make BVC deployment a 3-stage operation:
    1. new BVC()
    2. deposit()  // multiple times perhaps
    3. activate()

- Stop using a modifier for fundamental behavioral logic

- Reorganize activation conditions into two atomic modifiers
  (validation-only)

- Similarly, make the activate() function body behavior-only

gnidan commented:
OK, I believe I have addressed all your concerns @chatch. Please verify and let me know.

(There's a bunch of changes as you can see, tried to lay the commits out cleanly for your review)

I also realize now that there is another task related to this PR - making the BVCs transfer to WithdrawalReserves instead of the participant directly, in order to funnel the withdrawal operations into our planned infrastructure.

... but I'm going to make a separate ticket for that and not let this PR grow any larger.

chatch commented:
awesome. merging now.

hey how did you do that code peice by code peice referencing above? where it marks a snippet? useful!

gnidan commented:
> hey how did you do that code peice by code peice referencing above? where it marks a snippet? useful!

You can just browse "Files changed" at the top and click on a line to leave a comment. Super useful!

gnidan commented:
Thanks for the review!
